src/cmd/services/m3coordinator/ingest/carbon/ingest.go

```diff
@@ -148,9 +149,107 @@ type ingester struct {
 	metrics              carbonIngesterMetrics
 	tagOpts              models.TagOptions
 
+	lineResourcesPool pool.ObjectPool
+
+	sync.RWMutex
 	rules []ruleAndRegex
+}
 
-	lineResourcesPool pool.ObjectPool
+func (i *ingester) OnUpdate(clusterNamespaces m3.ClusterNamespaces) {
+	rules := i.regenerateIngestionRules(clusterNamespaces)
+	if rules == nil {
+		namespaces := make([]string, 0, len(clusterNamespaces))
+		for _, ns := range clusterNamespaces {
+			namespaces = append(namespaces, ns.NamespaceID().String())
+		}
+		i.logger.Warn("generated empty carbon ingestion rules from latest cluster namespaces update. leaving"+
+			" current set of rules as-is.", zap.Strings("namespaces", namespaces))
+		return
+	}
+
+	compiledRules, err := i.compileRules(*rules)
+	if err != nil {
+		i.logger.Error("failed to compile latest rules. continuing to use existing carbon ingestion "+
+			"rules", zap.Error(err))
+		return
+	}
+
+	i.Lock()
+	i.rules = compiledRules
+	i.Unlock()
+}
+
+func (i *ingester) regenerateIngestionRules(clusterNamespaces m3.ClusterNamespaces) *CarbonIngesterRules {
+	var (
+		rules = &CarbonIngesterRules{
+			Rules: i.opts.IngesterConfig.RulesOrDefault(clusterNamespaces),
+		}
```

Comment:
I think this whole section may need to be run under a lock since the Rules slice here from `i.opts.IngesterConfig.RulesOrDefault` can be mutated by the sort below

Reply:
Good catch. I updated this to just have the `OnUpdate` method acquire a lock when it's first called.